### PR TITLE
Lowering FORALL statement and FORALL construct with scalar, array, an…

### PR DESCRIPTION
**Pull request by schweitzpgi**
…d WHERE

construct assignments.

This patch includes changes to the bridge, array expression lowering, the array
value copy transformation, some FIR ops, and tests. Code generation was tested
end to end with the NAG F95 tests.

The implementation raises (nested) FORALL constructs (statements) into array
expression form. This means that array expressions and FORALL constructs will
be presented to the optimizer with the same structure. There are some notable
differences between FORALLs and array expressions that required extending the
array framework.

  - An array expression is isomorphic between the iteration space and the
    access expressions. FORALL need not be.
  - An array expression is limited by the standard to operate upon exactly
    one ranked component in a designator. A FORALL assignment is, generally,
    not limited in this way.
  - WHERE constructs contain array expressions both as masks and in the
    construct blocks. FORALL constructs may contain scalar assignments,
    array assignments, other FORALL constructs, and WHERE constructs.
  - WHERE constructs have a cascading disjunctional conditional execution
     structure over entire array assignments on an elemental basis. The if-else
     cascade is evaluated in program order. FORALL masks have a conjunctional
     conditional structure and do not, generally, have an elemental
     application but are a function of the user-defined iteration space. These
     two execution structures can be mixed with the FORALL case having
     priority.

Despite these differences, these Fortran constructs share the "copy-in
copy-out" execution model. See the standard for more information.

After the array copy value transformation runs, FIR code will have the same
fundamental structure for array assignments, WHERE constructs, FORALL
constructs, FORALL/WHERE constructs, and DO CONCURRENT loops. (That is,
the nominal "parallel" execution forms.)

This patch does not address FORALL and assignment to POINTERs/ALLOCATABLEs. It
also does not fix bugs in arrays of CHARACTER or derived types.

### CODE REVIEW

File: flang/lib/Lower/ConvertExpr.cpp
Comment (psteinfeld) [inline flang/lib/Lower/ConvertExpr.cpp:2884]: Did you mean "that describes it" rather than "the describes it"?

File: flang/lib/Lower/ConvertExpr.cpp
Comment (psteinfeld) [inline flang/lib/Lower/ConvertExpr.cpp:2934]: Did you mean "is defined" rather than "are defined"?

File: flang/lib/Lower/ConvertExpr.cpp
Comment (psteinfeld) [inline flang/lib/Lower/ConvertExpr.cpp:3284]: "Allocate a storage" reads awkwardly.  Did you mean to say "Allocate a storage temp"?

File: flang/lib/Lower/ConvertExpr.cpp
Comment (psteinfeld) [inline flang/lib/Lower/ConvertExpr.cpp:4589]: "as unbox value" reads awkwardly.  But I'm not sure what you meant.

File: flang/lib/Lower/SymbolMap.cpp
Comment (psteinfeld) [inline flang/lib/Lower/SymbolMap.cpp:306]: clang-format thinks that this line should move one space to the left.

**Review by psteinfeld — APPROVED** [review]
Thanks for doing this!  It looks like a huge effort.

I don't understand the code very much, but I did read through all of the text and made several nit comments.  You might want to wait for a review from someone more knowledgeable about lowering than I am.

Also, the description contains the spelling error "program otder" in the fourth bulleted point.

**Comment by schweitzpgi** [comment]
> Thanks for doing this! It looks like a huge effort.
> 
> I don't understand the code very much, but I did read through all of the text and made several nit comments. You might want to wait for a review from someone more knowledgeable about lowering than I am.
> 
> Also, the description contains the spelling error "program otder" in the fourth bulleted point.

Thanks for applying your eagle eye tool, Pete. I'll fix up these typos.

File: flang/lib/Lower/ConvertExpr.cpp
Comment (jeanPerier) [inline flang/lib/Lower/ConvertExpr.cpp:4364]: Does `explicitImpliedShape` necessarily relies on `Fortran::evaluate::GetShape` to be computed. I did not survey all the places where it is filled up, but it looks like it is mostly set with the result of `convertFEShape`, which I only works on constant shapes. Does this mean the raising only work with constant shape arrays for now ?

File: flang/include/flang/Lower/FIRBuilder.h
Comment (schweitzpgi) [reply flang/include/flang/Lower/FIRBuilder.h:257]: When we have a value of `box` type, that value must have been created with a `fir,shift` sort of op initially. That is, the origin shift information is included as part of the box. Therefore, in some sense, it may not be strictly necessary to pull the shift information out of the box and re-create a new ShiftOp to group that information. In one of the cases I was looking at, IIRC, there was an `embox` that was then used to create a new `shift` op.

I can remove this part of the comment. It's probably not that helpful.

File: flang/lib/Lower/ConvertExpr.cpp
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:3240]: I believe that the shape will be known and rectangular, yes.

There is one case where we can have ragged arrays, but that requires POINTER, additional dereferences, and different rather different IR to be correct. POINTER is still TODO here.

File: flang/lib/Lower/ConvertExpr.cpp
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:4244]: IIRC, the `AsGenericExpr` helper can fail to resolve to a valid template instance at compile time and spew a few pages of an error message.

File: flang/lib/Lower/ConvertExpr.cpp
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:4364]: Yes, that is a limitation for the moment.

File: flang/lib/Optimizer/Transforms/ArrayValueCopy.cpp
Comment (schweitzpgi) [reply flang/lib/Optimizer/Transforms/ArrayValueCopy.cpp:487]: There is some interplay going on here. `array_fetch` accepts an extended path through an array base value to a value in a recursive component. (That's for interference analysis.) `array_coor` does not allow this, but does abstract away applying custom origins, run-time extents, etc. like `array_load`. Here, `coordinate_of` is indexing into a particular element, `a(e)`, of an array that was captured with an `array_load`. That `array_load` captures runtime type parameters and we know that component arrays are rectangular and have well-defined shape in terms of those type parameters, so the trailing path can be lowered as pointer arithmetic, `a(e)%[path]`.

File: flang/lib/Lower/ConvertExpr.cpp
Comment (jeanPerier) [reply flang/lib/Lower/ConvertExpr.cpp:2304]: Thanks, that helps a lot.

File: flang/lib/Lower/ConvertExpr.cpp
Comment (jeanPerier) [reply flang/lib/Lower/ConvertExpr.cpp:2886]: Are you entirely sure that these expression have to be pure ? The standard states in 10.1.4 point 2 that:

> Execution of a function reference in [...] the concurrent-limits and concurrent-steps in a FORALL statement (10.2.4) is permitted to define variables in the subsidiary [...] forall-assignment-stmt.

That being said, if the FORALL concurrent-limits and concurrent-steps may depend on higher level FORALL indices, it sounds hard to ensure they are evaluated only once... So it may be worth studying a bit what is the behavior of existing compilers here before starting to do more complex things.

File: flang/lib/Lower/ConvertExpr.cpp
Comment (jeanPerier) [reply flang/lib/Lower/ConvertExpr.cpp:4304]: But it looks like [line 4342](https://github.com/schweitzpgi/f18-llvm-project/blob/79dd7cc4127fb4dffea14350a21c7833451c1248/flang/lib/Lower/ConvertExpr.cpp#L4342)  is only propagating an empty CC from the result of another `raiseToArray` call, which does not really explain how this empty CC was produced in the first place.

File: flang/lib/Optimizer/Transforms/ArrayValueCopy.cpp
Comment (jeanPerier) [reply flang/lib/Optimizer/Transforms/ArrayValueCopy.cpp:487]: Ok, so the handling of pointer, allocatable, and automatic components will be different in lowering. Thanks for the explanation.

**Review by jeanPerier — COMMENTED** [review]
Thanks for all the answers.

**Pull request merged**
